### ISSUE TITLE
Detect when sh cannot handle Windows short paths

### ISSUE THREAD

**Issue by acfoltzer** [haskell/cabal#3487]
**Summary by @ezyang:** In a misconfigured msys installation, the default sh in the path may come from cygwin which cannot handle Windows short paths. Unfortunately, Cabal 1.24 switched to using short paths, so using an sh which can handle this is important.

`./Setup configure` should verify that `sh` can handle short paths. It's not currently clear what the best way to do this is. It's also possible that cabal-install needs a check too.

Below is the original bug report.

----

I was having trouble with the configure step for `old-time`, and decided to compare the behavior of 1.22 and 1.24. 1.22 works fine, but it appears that while 1.24 successfully finds the correct path to gcc, it mangles it when setting `CC` and invoking the `configure` script.

```
...
Using gcc version 5.2.0 found on system at:
C:\Users\hudson\AppData\Local\Programs\stack\x86_64-windows\ghc-8.0.1\mingw\bin\gcc.exe
...
"C:\cygwin\bin\sh.exe" "./configure" "--with-compiler=ghc" "--prefix=C:\cygwin\tmp\old-time-1.1.0.3\.cabal-sandbox" "--bindir=$prefix\bin" "--libdir=$prefix" "--libexecdir=$prefix\$libname" "--datadir=$prefix" "--sysconfdir=$prefix\etc" "CC=C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe"
```

(full log at https://gist.github.com/acfoltzer/01fc8238acaafbb99e557f7320fcc318)

It seems that `x86_64-windows` is getting mangled into `X86_64~1`, and `ghc-8.0.1` into `GHC-80~1.1`. I'm honestly baffled; any thoughts?

**Comment by ezyang** [comment]
Those look like short path names to me. See b946c05e749fb0acb78ed8ae135b98e576df9d55

**Comment by 23Skidoo** [comment]
Yes, that's on purpose, to work around the issues that Autoconf has with paths that contain spaces. See #3185.

**Comment by acfoltzer** [comment]
From `config.log` for `old-time`:

```
configure:2256: checking for gcc
configure:2283: result: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe
configure:2512: checking for C compiler version
configure:2521: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe --version >&5
./configure: line 2523: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe: command not found
configure:2532: $? = 127
configure:2521: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe -v >&5
./configure: line 2523: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe: command not found
configure:2532: $? = 127
configure:2521: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe -V >&5
./configure: line 2523: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe: command not found
configure:2532: $? = 127
configure:2521: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe -qversion >&5
./configure: line 2523: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe: command not found
configure:2532: $? = 127
configure:2552: checking whether the C compiler works
configure:2574: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe -fno-stack-protector   conftest.c  >&5
./configure: line 2576: C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe: command not found
configure:2578: $? = 127
configure:2616: result: no
```

So cabal is taking a path that does exist and work, and is turning it into one that does not. Is that cabal's fault or is our Windows 10 system configured incorrectly?

**Comment by gbaz** [comment]
The question is why those paths do not work -- i.e. are they not valid short path names or is there a broader issue using short path names, i.e. maybe that the configure script itself can't handle short names for some reason?

**Comment by gbaz** [comment]
Here's a related problem I'm seeing -- as per https://github.com/haskell/cabal/pull/2840 we should be using the msys sh, even with cygwin. But somehow the cygwin sh is still being called even though I have "C:\Program Files\Haskell Platform\8.0.1\msys\usr\bin," set as an extra-prog-path in my cabal config file. Was there a regression here?

**Comment by gbaz** [comment]
Aha, I had

extra-prog-path: C:\Program Files\Haskell Platform\8.0.1\msys\usr\bin,
                          C:\Users\gbazerman\AppData\Roaming\cabal\bin

And apparently that format was silently bad.

If we remove the comma and the second element it works and builds for me. @acfoltzer try building with these three lines in your cabal config:

    extra-prog-path: C:\Program Files\Haskell Platform\8.0.1\msys\usr\bin
    extra-lib-dirs: C:\Program Files\Haskell Platform\8.0.1\mingw\lib
    extra-include-dirs: C:\Program Files\Haskell Platform\8.0.1\mingw\include

**Comment by acfoltzer** [comment]
The same problem arises, alas. It still tries to shorten the path, which on my system doesn't resolve correctly. I also tried with `network` with the same results, even with the usual `--configure-option --host=x86_64-unknown-mingw32` that I have to throw in for successful builds over a Cygwin ssh shell.

I should note that I'm only using Cygwin for the ssh capabilities; the path is configured so that the GHC-bundled mingw tools shadow Cygwin's.

**Comment by gbaz** [comment]
@acfoltzer I see in your paste above

    "C:\cygwin\bin\sh.exe" "./configure" "--with-compiler=ghc" "--prefix=C:\cygwin\tmp\old-time-1.1.0.3\.cabal-sandbox" "--bindir=$prefix\bin" "--libdir=$prefix" "--libexecdir=$prefix\$libname" "--datadir=$prefix" "--sysconfdir=$prefix\etc" "CC=C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe"

I.e. an invocation of the `sh` from cygwin. I got the error you described with that. When I ensured I was using the msys `sh` instead, then the build ran successfully....

**Comment by 23Skidoo** [comment]
Can confirm that using Cygwin tools can lead to breakage. We should perhaps add a warning about it...

**Comment by gbaz** [comment]
I like the idea of auto-detecting wrong tools and emitting a warning and a suggestion of how to configure things to use the right tools.

**Comment by acfoltzer** [comment]
Ah, I see what you mean now. I added the `/usr/bin` subdirectory of Stack's msys installation, and it appears to be using the right `sh.exe`, although with the same results:

```
"C:\Users\hudson\AppData\Local\Programs\stack\x86_64-windows\msys2-20150512\usr\bin\sh.exe" "./configure" "--with-compiler=ghc" "--prefix=C:\cygwin\tmp\old-time-1.1.0.3\.cabal-sandbox" "--bindir=$prefix\bin" "--libdir=$prefix" "--libexecdir=$prefix\$libname" "--datadir=$prefix" "--sysconfdir=$prefix\etc" "CC=C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe"
configure: WARNING: unrecognized options: --with-compiler
checking for gcc... C:\Users\hudson\AppData\Local\Programs\stack\X86_64~1\GHC-80~1.1\mingw\bin\gcc.exe
checking whether the C compiler works... no
```

I don't have graphical access to this machine, but I'll put in a ticket to install the new HP and see if the problem persists with that installation.

**Comment by acfoltzer** [comment]
Okay, using the Haskell Platform and with the suggested modifications to the cabal config, this is now resolved for me. It appears that the short path is still being passed to `sh`, but the `sh` bundled with the platform seems to agree with it. I'm fine with closing this ticket unless y'all think there might be some improvements to be made to help people reach a solution.

**Comment by 23Skidoo** [comment]
So I can think of two enhancements:

1. Shortening path names only when they contain spaces.
1. Detecting broken versions of `sh` and printing a warning.

**Comment by gbaz** [comment]
I don't think the former is very wise, as users should be encouraged to have a proper version of sh lying around (and indeed one comes with the msys tools). The latter seems like a great idea.

**Comment by FPtje** [comment]
I had tried installing msys2 (it was using Git's sh before), the cabal options described by @gbaz and nothing worked. When I changed my `PATH` back to ghc-7.10.3 everything worked again. Then change to `ghc-8.0.1` and everything breaks again.

My solution is even simpler: *something*, whether it be cabal, ghc, sh or whatever cannot deal with `.` or `-` in folder names. Change `ghc-8.0.1` to something else (I changed it to just `8`), and this incompetence will no longer bother you.

Make sure to change `PATH` accordingly.

I should not have to go through this effort to get a simple compiler to work.
